File: Server/src/main/java/com/mancersoft/litevpnserver/transport/UdpTransport.java
package com.mancersoft.litevpnserver.transport;

import com.mancersoft.litevpnserver.ConnectionParams;
import trikita.log.Log;

import java.net.InetSocketAddress;
import java.net.SocketAddress;
import java.net.StandardProtocolFamily;
import java.net.StandardSocketOptions;
import java.nio.ByteBuffer;
import java.nio.channels.DatagramChannel;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import static com.mancersoft.litevpnserver.VpnManager.*;

public class UdpTransport implements IVpnTransport {

    private DatagramChannel mUdpChannel;
    private final String mSecret;
    private final int mPort;
    private final ConnectionParams mParams;

    private final ExecutorService mExecutorService = Executors.newCachedThreadPool();

    public UdpTransport(String secret, int port, ConnectionParams params) {
        mSecret = secret;
        mPort = port;
        mParams = params;
    }

    @Override
    public boolean start() {
        try {
            mUdpChannel = DatagramChannel.open(StandardProtocolFamily.INET6);
            mUdpChannel.setOption(StandardSocketOptions.SO_REUSEADDR, true);
            mUdpChannel.configureBlocking(true);
            var address = new InetSocketAddress(mPort);
            mUdpChannel.bind(address);
        } catch (Exception e) {
            Log.e(TAG, "UdpTransport start error", e);
            return false;
        }

        return true;
    }

    @Override
    public void stop() {
        try {
            if (mUdpChannel != null) {
                mUdpChannel.close();
            }
        } catch (Exception e) {
            Log.e(TAG, "UdpTransport stop error", e);
        }
    }

    @Override
    public boolean receiveConnection(Packet packet, String vpnIpAddress) {
        try {
            String receivedSecret = new String(packet.getData(), 1, packet.getLength() - 1);
            boolean isCorrectSecret = mSecret.equals(receivedSecret);
            if (isCorrectSecret) {
                mParams.setAddress(vpnIpAddress, mParams.getAddressPrefixLength());
                byte[] parameters = mParams.getBytes();
                parameters[PACKET_TYPE_BYTE_OFFSET] = CONNECT_PACKET;
                var paramPacket = new Packet();
                paramPacket.setData(parameters);
                paramPacket.setLength(parameters.length);
                paramPacket.setDestination(packet.getSource());
                for (int i = 0; i < 3; ++i) {
                    this.sendAsync(paramPacket);
                }
            }

            return isCorrectSecret;
        } catch (Exception e) {
            Log.e(TAG, "UdpTransport receiveConnection error", e);
            return false;
        }
    }

    @Override
    public void sendAsync(Packet packet) {
        mExecutorService.execute(() -> {
            try {
                var buffer = ByteBuffer.wrap(packet.getData());
                buffer = buffer.limit(packet.getLength());
                mUdpChannel.send(buffer, (SocketAddress) packet.getDestination());
            } catch (Exception e) {
                Log.e(TAG, "UdpTransport sendAsync error", e);
            }
        });
    }

    @Override
    public void receive(Packet packet) {
        try {
            var buffer = ByteBuffer.wrap(packet.getData());
            packet.setSource(mUdpChannel.receive(buffer));
            packet.setLength(buffer.position());
        } catch (Exception e) {
            Log.e(TAG, "UdpTransport receive error", e);
        }
    }
}

File: Server/src/main/kotlin/com/mancersoft/litevpnserver/transport/UdpTransport.kt
package com.mancersoft.litevpnserver.transport

import com.mancersoft.litevpnserver.VpnManager
import trikita.log.Log
import java.net.InetSocketAddress
import java.net.SocketAddress
import java.net.StandardProtocolFamily
import java.net.StandardSocketOptions
import java.nio.ByteBuffer
import java.nio.channels.ClosedByInterruptException
import java.nio.channels.DatagramChannel
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executors
import java.util.concurrent.Future

class UdpTransport(private val mPort: Int) : IVpnTransport {
    private var mUdpChannel: DatagramChannel? = null
    private val mExecutorService = Executors.newCachedThreadPool()
    private var mIncomingMessageProcessing: Future<*>? = null
    private var mMessageListener: ((Packet) -> Unit)? = null
    override val isReliable: Boolean
        get() = false

    override fun start(): CompletableFuture<Boolean> {
        val result = CompletableFuture<Boolean>()
        try {
            stop()
            mUdpChannel = try {
                DatagramChannel.open(StandardProtocolFamily.INET6)
            } catch (e: UnsupportedOperationException) {
                DatagramChannel.open(StandardProtocolFamily.INET)
            }

            mUdpChannel!!.setOption(StandardSocketOptions.SO_REUSEADDR, true)
            mUdpChannel!!.configureBlocking(true)
            val address = InetSocketAddress(mPort)
            mUdpChannel!!.bind(address)
            mIncomingMessageProcessing = mExecutorService.submit { incomingMessagesProcessing() }
        } catch (e: Exception) {
            Log.e(VpnManager.TAG, "UdpTransport start error", e)
            result.complete(false)
            return result
        }
        result.complete(true)
        return result
    }

    override fun stop() {
        try {
            mIncomingMessageProcessing?.cancel(true)
            mUdpChannel?.close()
        } catch (e: Exception) {
            Log.e(VpnManager.TAG, "UdpTransport stop error", e)
        }
    }

    override fun sendAsync(packet: Packet) {
        val packetToSend = Packet(packet)
        mExecutorService.execute {
            try {
                val buffer = ByteBuffer.wrap(packetToSend.data, 0, packetToSend.length)
                mUdpChannel!!.send(buffer, packetToSend.destination as SocketAddress)
            } catch (e: Exception) {
                Log.e(VpnManager.TAG, "UdpTransport sendAsync error", e)
            }
        }
    }

    override fun setOnMessageListener(messageListener: ((Packet) -> Unit)?) {
        mMessageListener = messageListener
    }

    private fun incomingMessagesProcessing() {
        try {
            val packet = Packet()
            packet.data = ByteArray(VpnManager.MAX_PACKET_SIZE)
            while (!Thread.currentThread().isInterrupted) {
                //System.out.println("Start receive transport");
                receive(packet)
                //System.out.println("Stop receive transport");
                mMessageListener?.invoke(Packet(packet))
            }
        } catch (e: Exception) {
            Log.e(VpnManager.TAG, "UdpTransport incomingMessagesProcessing error", e)
            restart()
        }
    }

    private fun restart() {
        Log.d(VpnManager.TAG, "UdpTransport restarting...")
        start()
    }

    private fun receive(packet: Packet) {
        try {
            val buffer = ByteBuffer.wrap(packet.data)
            packet.source = mUdpChannel!!.receive(buffer)
            packet.length = buffer.position()
        } catch (ignored: ClosedByInterruptException) {
            Log.d(VpnManager.TAG, "UdpTransport receive interrupted")
        } catch (e: Exception) {
            Log.e(VpnManager.TAG, "UdpTransport receive error", e)
        }
    }

}